workspaces/liferay-dlfoldertemplate-workspace/client-extensions/liferay-dlfoldertemplate-custom-element/src/components/template-diagram/Diagram.js

```diff
@@ -0,0 +1,505 @@
+	};
+
+	const updateDiagramDataSourceLocally = useCallback(
+		(currentNodes, currentEdges, idsToExclude) => {
+			const filteredNodes = currentNodes.filter(
+				(node) => !idsToExclude.includes(node.id)
+			);
+
+			const filteredEdges = currentEdges.filter(
+				(edge) => !idsToExclude.includes(edge.source)
+			);
+
+			const {
+				edges: layoutedEdges,
+				nodes: layoutedNodes,
+			} = getLayoutedElements(filteredNodes, filteredEdges);
+
+			setNodes([...layoutedNodes]);
+
+			setEdges([...layoutedEdges]);
+		},
+		[setNodes, setEdges]
+	);
+
+	const updateDiagramSingleNodeLocally = useCallback(
+		(updatedNode) => {
+			const selectedDiagramNode = nodes.filter(
+				(node) => node.id.toString() === updatedNode.id
+			);
+
+			if (selectedDiagramNode && !!selectedDiagramNode.length) {
+				selectedDiagramNode[0].data.description =
+					updatedNode.description;
+				selectedDiagramNode[0].data.label = updatedNode.name;
+			}
+		},
+		[nodes]
+	);
+
+	const onAdd = useCallback(
+		(parentNodeId) => {
+			const addNodeAsync = async () => {
+				const {
+					edges: layoutedEdges,
+					nodes: layoutedNodes,
+				} = await addNewNode(parentNodeId, templateId, nodes, edges);
+
+				setNodes([...layoutedNodes]);
+
+				setEdges([...layoutedEdges]);
+			};
+
+			addNodeAsync();
+		},
+		[nodes, edges, setNodes, setEdges, templateId]
+	);
+
+	const onDelete = useCallback(
+		(params) => {
+			try {
+				const nodeId = params[0].id || params || selectedNode.id;
+
+				const nodesToBeDeleted = getSubNodes(nodeId);
+
+				updateDiagramDataSourceLocally(nodes, edges, nodesToBeDeleted);
+
+				deleteNodes(nodesToBeDeleted);
+
+				setSelectedNode(null);
+			}
+			catch (error) {
+				showError(error.message);
+			}
+		},
+		[nodes, edges, updateDiagramDataSourceLocally, selectedNode]
+	);
+
+	const loadNodes = useCallback(
+		async (templateId) => {
+			const {
+				edges: layoutedEdges,
+				nodes: layoutedNodes,
+			} = await getNodesFromHeadless(templateId);
+
+			setNodes([...layoutedNodes]);
+
+			setEdges([...layoutedEdges]);
+		},
+		[setEdges, setNodes]
+	);
+
+	useEffect(() => {
+		loadNodes(templateId);
+	}, [templateId, loadNodes]);
+
+	const onSave = () => {
+		form.validateFields()
+			.then(
+				async (values) => {
+					try {
+						setIsLoading(true);
+
+						await updateFolderTemplate(values.id, values);
+
+						setIsLoading(false);
+
+						updateDiagramSingleNodeLocally(values);
+					}
+					catch (error) {
+						setIsLoading(false);
+
+						showError(error.message);
+					}
+				},
+				(error) => {
+					showError(error.message);
+				}
+			)
+			.catch((error) => {
+				showError(error.message);
+			});
+	};
+
+	return (
+		<>
+			{nodes && edges && (
+				<ReactFlow
+					connectionLineType={ConnectionLineType.SmoothStep}
+					edges={edges}
+					fitView
+					key={key}
+					nodeTypes={{
+						folderNode: (props) => (
+							<FolderNode
+								{...props}
+								onAdd={onAdd}
+								onDelete={onDelete}
```

Comment:
usually we would name this like this: `onDelete={handleDelete}`. Basically, `onDelete` calls the `handleDelete` function. The only exception to this would be if `onDelete` we were directly calling a function that is passed in as a prop to the current component. Then it would be okay for onDelete, to call on onDelete, which likely in the parent component would be calling handleDelete.

Reply:
onAdd is the same situation... The function that gets called should be `handleAdd`.

Reply:
Check for any others.